### PR TITLE
tsh: tctl: Add basic session ls functionality

### LINKED ISSUE #11205: Allow a way for `tsh` to be able to fetch a list of session recordings available for play
### What
Feature request is to allow the `tsh` tool to list session recordings available for playback to an end user.

### How
Something like `tsh sessions ls` to list out all session recording ids compatible with `tsh play` over a specified timeframe or most recent entries.

- [ ] By default `tsh sessions ls` should output table with all sessions containing columns: session ID, session type (can we?), session timestamp, participants, server information.
- [ ] The command should support `--format` flag to print sessions in JSON/YAML formats.
- [ ] Command should use pagination for events API.
- [ ] Command should allow to fetch sessions of particular type e.g. SSH vs Kubernetes:
  * `tsh sessions ls --type=[ssh|k8s]`.
- [ ] Command should allow to fetch sessions for particular time interval: 
  * `tsh sessions ls --from-utc=xxx --to-utc=xxx`
- [ ] Confirm that after listing sessions with `tsh sessions ls` you should be able to use `tsh play` to play the sesion.

### Implementation

Use events API to fetch `session.end` events. Use pagination.

### Why
This would be a big quality of life for users to not have to grab the session id from the Web UI before reviewing a session recording.

### Workaround
Get the session id from the web UI or audit logging

### PR DESCRIPTION
**Pull request by EdwardDowling**
example output

```
[edward@edward-21a0003xus teleport]$ build/tsh recordings ls
ID                                   Type Participants   Hostname          Timestamp           
------------------------------------ ---- -------------- ----------------- ------------------- 
958e80e1-1fc9-4f31-ab56-363a1f0a99d0      edward         edward-21a0003xus Jun  3 10:20:14 UTC 
eda81a2e-6a4d-4c4b-aa93-f5a2fb3eaeb3 kube edward                           Jun  3 10:18:16 UTC 
5e030636-76fa-408c-97db-7732109b4d91 kube edward                           Jun  3 10:08:26 UTC 
e202e908-6a3f-42ab-87aa-cd76a67be587      edward, edward edward-21a0003xus Jun  3 09:44:27 UTC 
f2a62a49-2541-4eeb-a5ef-4f3eaad69fa6 kube edward                           Jun  2 10:18:14 UTC 
ca94283e-78cd-4a72-8d5a-5a3b1d612b11 kube edward                           Jun  2 10:16:19 UTC
```

```
[edward@edward-21a0003xus teleport]$ build/tsh recordings ls --from-utc="2022-07-06T15:30:05Z"   --limit=3
ID                                   Type Participants Hostname          Timestamp           
------------------------------------ ---- ------------ ----------------- ------------------- 
98d2b9f7-432a-42d9-989e-b9932e088521      edward       edward-21a0003xus Jul  6 16:08:13 UTC 
db37643b-c9e7-4788-b80a-6e47019cf28a      edward       edward-21a0003xus Jul  6 17:49:36 UTC 
98e145f3-bd62-41a2-a1b3-e10d8a9ed0e8      edward       edward-21a0003xus Jul  6 17:50:03 UTC
```

closes #11205

### CODE REVIEW

File: tool/tctl/common/sessions_command.go
Comment (r0mant) [inline tool/tctl/common/sessions_command.go:84]: To make the error message more useful:

1. Use `trace.BadParameter` instead of `trace.Errorf`.
2. Include actual value we failed to parse.
3. Say what the expected format is.

In general, we should aim to make our messages look like proper messages, we have a tendency of returning errors that read more like internal debug logs. For example, in this case we should say:

```
Failed to parse session listing start time: expected format XXX, got YYY.
```

File: tool/tctl/common/sessions_command.go
Comment (r0mant) [inline tool/tctl/common/sessions_command.go:90]: Same comment as above about the error message.

File: tool/tctl/common/sessions_command.go
Comment (r0mant) [inline tool/tctl/common/sessions_command.go:116]:
```suggestion
	case teleport.Text, "":
```

File: tool/tctl/common/sessions_command.go
Comment (nklaassen) [inline tool/tctl/common/sessions_command.go:100]: pagination is duplicated in tsh and tctl, you could probably pull that into lib/client

File: tool/tctl/main.go
Comment (r0mant) [inline tool/tctl/main.go:45]: I think you need to add this to enterprise's tctl also: https://github.com/gravitational/teleport.e/blob/master/tool/tctl/main.go.

File: tool/tctl/common/sessions_command.go
Comment (r0mant) [inline tool/tctl/common/sessions_command.go:56]: Nit: Do we have a list of supported formats somewhere in a variable so we don't have to enumerate them here? In tsh we have `defaultFormats`, can we move it some commonplace it can be reused from?

File: tool/tsh/tsh.go
Comment (r0mant) [inline tool/tsh/tsh.go:572]: Same comments as I left for tctl's new command apply here too: --limit, better descriptions, defaults.

File: tool/tsh/tsh.go
Comment (r0mant) [inline tool/tsh/tsh.go:2068]: You have exactly the same function in tctl, can we reuse?

File: tool/tctl/common/sessions_command.go
Comment (r0mant) [inline tool/tctl/common/sessions_command.go:62]: Move to lib/defaults.

File: tool/tsh/tsh.go
Comment (r0mant) [inline tool/tsh/tsh.go:3785]: Move to lib/defaults and use in both places.

File: tool/tsh/tsh.go
Comment (r0mant) [inline tool/tsh/tsh.go:3791]: A lot of this logic is the same between tctl and tsh, let's factor it out and reuse to avoid duplication.

**Review by r0mant — CHANGES_REQUESTED** [review]

File: tool/tctl/common/sessions_command.go
Comment (nklaassen) [reply tool/tctl/common/sessions_command.go:84]: agreed, except that errors should start with a lower case character and should not end with a period to match the rest of the codebase, when the error is logged the sentence format may not make sense

**Review by nklaassen — COMMENTED** [review]
agreed with Roman's review

File: tool/tctl/common/collection.go
Comment (EdwardDowling) [reply tool/tctl/common/collection.go]: sgtm

File: tool/tctl/main.go
Comment (EdwardDowling) [reply tool/tctl/main.go:45]: Should that be done in this PR or should I make a separate PR for the change to the enterprise submodule

File: tool/tctl/common/sessions_command.go
Comment (EdwardDowling) [reply tool/tctl/common/sessions_command.go]: With the addition of the --limit pagination on top of that doesn't make sense to me. Now users can just specify the start time/endtime and how many they need.

File: lib/defaults/defaults.go
Comment (zmb3) [inline lib/defaults/defaults.go:771]: It's not clear to me that this is for tsh/tctl. Can you think of a better name or better place to put this constant?

File: lib/client/api.go
Comment (zmb3) [inline lib/client/api.go:4314]: Do you need to close the authClient too?

File: tool/tctl/common/sessions_command.go
Comment (zmb3) [inline tool/tctl/common/sessions_command.go:62]: What does per-page mean in this context? Are you implementing paging in the CLI?

File: tool/tsh/tsh.go
Comment (zmb3) [inline tool/tsh/tsh.go:572]: Any reason the help message for `tctl sessions` is different from `tsh sessions`?

File: tool/tsh/tsh_test.go
Comment (zmb3) [inline tool/tsh/tsh_test.go:2753]: Where is the `showSessions` func that this is calling?

File: tool/tsh/tsh_test.go
Comment (zmb3) [inline tool/tsh/tsh_test.go:2743]: Please also add a test for `events.WindowsDesktopSessionEnd`.

File: tool/tsh/tsh.go
Comment (EdwardDowling) [reply tool/tsh/tsh.go:572]: Fixed

File: tool/tctl/common/sessions_command.go
Comment (EdwardDowling) [reply tool/tctl/common/sessions_command.go]: Ah you are right.
This doesn't really make much sense as a message anymore now that the limiting is done through the --limit

File: lib/defaults/defaults.go
Comment (EdwardDowling) [reply lib/defaults/defaults.go]: how about defaults.TshTctlSessionListLimit

File: tool/tctl/common/sessions_command.go
Comment (nklaassen) [inline tool/tctl/common/sessions_command.go:83]:
```suggestion
		return trace.Errorf("getting session events: %v", err)
```

File: tool/tctl/common/sessions_command.go
Comment (nklaassen) [inline tool/tctl/common/sessions_command.go:80]: there is no pagination, what if `maxSessionsToShow` is too large to fit in a single page? I think you can support both pagination and a global limit

File: lib/defaults/defaults.go
Comment (nklaassen) [reply lib/defaults/defaults.go]: I'm not sure about introducing `tool/common`, it kind of implies there's a tool called `common`. I'd lean towards somwhere in `lib` for shared code, probably under `lib/client`

**Review by nklaassen — CHANGES_REQUESTED** [review]

File: lib/defaults/defaults.go
Comment (EdwardDowling) [reply lib/defaults/defaults.go]: I've moved all the shared code to lib/client/tool_utils.go

File: tool/tctl/common/sessions_command.go
Comment (EdwardDowling) [reply tool/tctl/common/sessions_command.go]: I've added back the pagination.

**Comment by EdwardDowling** [comment]
@r0mant @zmb3 @nklaassen could I get a re-review on this?

File: lib/defaults/defaults.go
Comment (nklaassen) [inline lib/defaults/defaults.go:774]: would `(api/defaults).DefaultChunkSize` be appropriate?

**Review by nklaassen — COMMENTED** [review]
Getting there, I just think the pagination could be a bit cleaner

File: lib/client/tool_utils.go
Comment (r0mant) [inline lib/client/tool_utils.go:61]: I think this belongs in lib/defaults.

File: lib/client/tool_utils.go
Comment (r0mant) [inline lib/client/tool_utils.go:84]: I'm not sure this belongs in lib/client TBH. Can we make a "common" subpackage in tool and put it there instead?

**Review by r0mant — CHANGES_REQUESTED** [review]
I think we should take apart `tool_utils.go` and put things in the appropriate places instead. Dumping unrelated things in a generic "util" module is rarely a good idea.

File: lib/client/tool_utils.go
Comment (nklaassen) [reply lib/client/tool_utils.go:84]: This was already suggested, he made the change, then I suggested the opposite https://github.com/gravitational/teleport/pull/12763#discussion_r916110339 🤦‍♂️ Sorry Edward. I'll defer to Roman on this one

**Comment by EdwardDowling** [comment]
> I realize this is just reusing the existing api to retrieve session, but would this all be much simpler and performant if we added a new streaming api? It would remove the need to limit and paginate sessions which would make this less complex and require less roundtrips which would be more performant
@rosstimothy was this streaming api a suggestion for this PR or to gauge the usefulness of such an api in the future?

**Review by r0mant — APPROVED** [review]

File: lib/defaults/defaults.go
Comment (zmb3) [inline lib/defaults/defaults.go:773]: Nit: it's odd to see these tools written in all caps.
```suggestion
// Default values for tsh and tctl commands.
```

File: lib/defaults/defaults.go
Comment (zmb3) [inline lib/defaults/defaults.go:786]: Wouldn't hurt to add a couple of tests for this.

File: tool/common/common.go
Comment (zmb3) [inline tool/common/common.go:43]: If the backend is dynamo and we don't specify a session ID, this will make a separate query for each day in the range, which can really hammer the backend with requests.

Should we enforce some maximum number of days you're allowed to search? If not, someone can specify a time of 0 and put a lot of stress on the system to make all of those queries.

Comment (EdwardDowling) [reply tool/common/common.go:43]: I can add a limit if you think that would make a better experience, but since the default range is one day in the past and to have more the user would need to explicitly ask for a wider range i'm not sure preventing them would be a very pleasant user experience. 
Maybe a warning would be better?

**Comment by EdwardDowling** [comment]
@zmb3 Could you give this another once over when you get a second?

**Comment by zmb3** [comment]
As discussed in a separate issue, Edward is going to have a look at RFD19 and determine whether there are any changes here as a result of that information.

File: tool/common/common.go
Comment (zmb3) [reply tool/common/common.go:43]: I think we probably need some sort of upper limit.

The issue is if they request a very large range, Teleport will continue to make a query for each day, _even if they ctrl+c and abandon the request_.

Comment (EdwardDowling) [reply tool/common/common.go:43]: I have added in a limit of 365 days.

**Comment by r0mant** [comment]
@nklaassen You requested changes here before, could you please take another look?

**Comment by EdwardDowling** [comment]
@nklaassen could you have another look at this when you get a second?

File: lib/client/api.go
Comment (nklaassen) [inline lib/client/api.go:74]: I've gotta say, I'm pretty uncomfortable with `lib` calling out `tool`. Generally we have `tool/<tool_name>` containing all the code specific to that tool. Common code shared between tools is always under `lib` or `api`. I'm okay with adding `tool/common` for shared code that is close to the CLI (like `WriteJSON`/`WriteYAML` helpers,) but making `lib` depend on `tool` seems quite backwards to the rest of the dependency graph.

`tool/common` also depends on other packages under `lib`, so someone is bound to make an innocuous change in `lib` that creates a dependency cycle, and this could get tricky to untangle. IMO `GetPaginatedSessions` should really be in `lib/client` and that would solve this.

@r0mant what do you think? @EdwardDowling sorry to keep suggesting moving things around.

File: tool/common/common.go
Comment (nklaassen) [inline tool/common/common.go:43]: No need to request 50 if we only need 1.
```suggestion
		if remaining := max - len(sessions); remaining > pageSize {
			pageSize = remaining
		}
		nextEvents, eventKey, err := authClient.SearchSessionEvents(fromUTC, toUTC,
			pageSize, order, prevEventKey, nil /* where condition */, "" /* session ID */)
```

File: tool/common/common.go
Comment (nklaassen) [inline tool/common/common.go:48]:
```suggestion
		if eventKey == "" || len(sessions) >= max {
```

File: lib/client/api.go
Comment (EdwardDowling) [reply lib/client/api.go]: Yeah this makes sense, will move it now.

**Review by rosstimothy — APPROVED** [review]

**Review by nklaassen — APPROVED** [review]
Code looks good.

I notice the e submodule ref is updated here, but the referenced commit doesn't seem to include any changes to add this command to tctl. Make sure you update that correctly, do you have a corresponding PR in teleport.e?

**Comment by EdwardDowling** [comment]
> Code looks good.
> 
> I notice the e submodule ref is updated here, but the referenced commit doesn't seem to include any changes to add this command to tctl. Make sure you update that correctly, do you have a corresponding PR in teleport.e?

Fixed the submodule ref there.
Will be making a seperate PR to e once this is merged

**Review by xacrimon — CHANGES_REQUESTED** [review]
This looks neat, but can we revise the command naming or at least discuss it? `tsh kube` has a subcommand called `sessions` for listing *active sessions* and if we call this `sessions` while it lists recordings, that seems quite confusing to me. I think we also have an open feature request to add active session listing for SSH sessions and we'd get a name collision there. What about something like `tsh recordings ls`?

**Comment by EdwardDowling** [comment]
> This looks neat, but can we revise the command naming or at least discuss it? `tsh kube` has a subcommand called `sessions` for listing _active sessions_ and if we call this `sessions` while it lists recordings, that seems quite confusing to me. I think we also have an open feature request to add active session listing for SSH sessions and we'd get a name collision there. What about something like `tsh recordings ls`?

Recordings sounds like a good name to me. I've changed it there.

**Review by xacrimon — APPROVED** [review]

**Pull request merged**
